src/dune_file_watcher/dune_file_watcher.ml

```diff
@@ -640,28 +690,40 @@ let add_watch t path =
         loop ext
       in
       match ext with
-      | None -> Ok ()
+      | None -> ()
       | Some ext -> (
         let watch =
           lazy
             (fsevents ~latency:f.latency f.scheduler ~paths:[ path ]
                fsevents_standard_event)
         in
         match Watch_trie.add f.external_ ext watch with
-        | Watch_trie.Under_existing_node -> Ok ()
+        | Watch_trie.Under_existing_node -> ()
         | Inserted { new_t; removed } ->
           let watch = Lazy.force watch in
           Fsevents.start watch f.runloop;
           List.iter removed ~f:(fun (_, fs) -> Fsevents.stop fs);
-          f.external_ <- new_t;
-          Ok ())))
+          f.external_ <- new_t)))
   | Fswatch _ ->
     (* Here we assume that the path is already being watched because the coarse
        file watchers are expected to watch all the source files from the
        start *)
-    Ok ()
-  | Inotify inotify -> (
-    try Ok (Inotify_lib.add inotify (Path.to_string path)) with
-    | Unix.Unix_error (ENOENT, _, _) -> Error `Does_not_exist)
+    ()
+  | Inotify { inotify; awaiting_creation; mutex } ->
+    Mutex.lock mutex;
```

Comment:
Same concern here: messing around with `mutex` for every watch, even though most watches don't care about the `awaiting_creation` logic, seems wrong.